### DISCUSSION TITLE
Local deployment

### DISCUSSION THREAD

**Discussion by Tomaubier** [oleeskild/obsidian-digital-garden#160] [Q&A] [answered]
As a PhD student relying heavily on Obsidian, I'm looking for a way to share some of my notes with my lab colleagues.
This project seems to fit my needs perfectly however I'm wondering wether it would be possible to perform the deployment on a local server on a docker container for example?
Best regards
Tom

**Comment by oleeskild** [comment, marked as answer]
I don't know your skill level when it comes to node and docker, so excuse me if I'm either over or under explaining this. Let me know if you have any questios, or something isn't clear.

I threw something together you could use to run this locally:

1. Clone your digital garden repo to your computer
2. run `npm install express --save` to install express as a dependency.
3. Create a new file in the root folder called app.js, and paste the following code:

```js
const express = require('express')
const app = express()
const port = 8080; 

const searchHandler = require('./netlify/functions/search/search.js').handler;

app.get('/api/search', async (req, res) => {
  //Mock the netlify event
  let event = {queryStringParameters: req.query};

  let response = await searchHandler(event);
  res.status(response.statusCode).send(response.body);
})

app.use(express.static('dist'))

app.get('*', (req, res)=>{
  res.redirect("/404")
})

app.listen(port, () => {
  console.log(`Digital garden running on port ${port}`)
})
```
4. Run `npm install` 
5. Run `npm run build`
6. Run `node app.js`

The site should now be available at localhost:8080.
 
You can then use something like [pm2](https://pm2.keymetrics.io/) to run it on your server, and use a r[everse nginx proxy with lets encrypt](https://seanthegeek.net/1035/how-to-configure-a-nginx-reverse-proxy-with-lets-encrypt-certificates/) to expose it on port 443 with https. 

The one thing it is missing is automatically pulling the latest changes when you publish a note.

I will make a docker file available that does all this sometime in the future, when I find the time, so everybody has this option easily available.

Hopefully this is enough to get you started. 

If you manage to actually create a docker file for this, I would be very happy to accept a PR 😄

**Reply by Tomaubier** [reply]
Awesome!! Thanks for the quick and very detailed response!
I'm afraid I won't be able to achieve the docker setup, this isn't my area of expertise at all! 😅 I will try though! I will keep you updated!

One last question (which might be a bit more challenging to implement): Would it be possible to link the Digital Garden Obsidian plugin with any git repositories (located on a local private git server for exemple) by providing its url? Are some of Github features mandatory for this to work?

Thanks again for the help!
Best
Tom

**Reply by oleeskild** [reply]
Unfortunately, no. You need to use github, as the plugin is using the github API, not git cli commands, so that it wasn't dependent on users having git installed. My goal with this plugin was that it should be easy to use even if you'd never heard about git before, so that's why I opted for using the API instead.
Maybe some day down the line it would be possible to create a GitLab integration or something as well, that could be used on private servers. But currently this isn't something that I'm prioritizing.

**Reply by stories-with-dice** [reply]
Hey! I'm trying to use this but I'm getting some errors

when I did `build` I saw

```
npm run build

> web@1.0.0 build
> npm-run-all get-theme build:*


> web@1.0.0 get-theme
> node src/site/get-theme.js

/home/cristian/Downloads/digitalgarden-main/node_modules/path-scurry/dist/cjs/index.js:352
    #resolveParts(dirParts) {
                 ^

SyntaxError: Unexpected token '('
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/home/cristian/Downloads/digitalgarden-main/node_modules/glob/dist/cjs/glob.js:5:23)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
ERROR: "get-theme" exited with 1.
```

then when I do `node app.js` in Firefox I get

```
The page isn’t redirecting properly

Firefox has detected that the server is redirecting the request for this address in a way that will never complete.

    This problem can sometimes be caused by disabling or refusing to accept cookies.
```

**Reply by ooker777** [reply]
When I try doing this with the doc itself, I get this error: [`Having trouble rendering njk template ./src/site/feed.njk`](https://github.com/oleeskild/digitalgardendocs/issues/48)

**Reply by davidkopp** [reply]
My current relatively simple self-hosted setup using nginx, podman and a script that automatically creates a new container when changes are made...

Dockerfile:
```
FROM node:18 as build
WORKDIR /usr/src/app
COPY package*.json ./
RUN npm install
COPY . .
RUN npm run build

FROM node:18-slim
RUN apt-get update && apt-get install -y --no-install-recommends dumb-init
ENV NODE_ENV production
WORKDIR /usr/src/app

COPY package*.json ./
RUN npm ci --omit=dev
RUN npm install express --save

USER node
COPY --chown=node:node --from=build /usr/src/app/server.js ./
COPY --chown=node:node --from=build /usr/src/app/dist/ ./dist/

EXPOSE 8080
CMD ["dumb-init", "node", "server.js"]
```

The file `server.js` equals the file `app.js` from https://github.com/oleeskild/obsidian-digital-garden/discussions/160#discussioncomment-4671668 with one modification:
Removing line `const searchHandler = require('./netlify/functions/search/search.js').handler;`

Script (`~/deploy-digital-garden.sh`):
```sh
#!/bin/sh -eu

cd ~/my-digital-garden
git fetch origin
if git status | grep -q behind; then
  git merge origin/main
  podman build . -t my-digital-garden
  podman ps -q --filter "name=my-digital-garden" | grep -q . && podman stop my-digital-garden && podman rm -fv my-digital-garden
  podman run --name my-digital-garden -p 8080:8080 -d my-digital-garden
fi
```

Cronjob (` crontab -e`, triggers every 5 minute):

```crontab
*/5 * * * * ~/deploy-digital-garden.sh
```

Nginx reverse proxy (excerpt):
```
upstream mydigitalgarden {
  server localhost:8080;
}

server {
  server_name ...
  location / {
    proxy_pass http://mydigitalgarden;
  }
  ...
}
```

If anyone has suggestions for improvements, let me know.

**Reply by davidkopp** [reply]
Alternative Dockerfile using Nginx:

```Dockerfile
FROM node:18 as base
WORKDIR /usr/src/app
COPY package*.json ./
RUN npm install
COPY . .

FROM base as builder
WORKDIR /usr/src/app
RUN npm run build

FROM nginx:1.24
COPY ./nginx/nginx.conf /etc/nginx/conf.d/default.conf
COPY --from=builder /usr/src/app/dist/ /usr/share/nginx/html/
EXPOSE 8080
```

Nginx config file (`nginx/nginx.conf`):

```
server { 

    listen 8080;
    server_name localhost;

    error_page 404 /404/index.html;
    charset UTF-8;

    root /usr/share/nginx/html;

    location / {
        index index.html
        try_files $uri $uri/ =404;
    }
}
```

@oleeskild Do you know of any drawback of using Nginx instead of Express?
What is the purpose of the API endpoint `/api/search` in your `app.js` file?

**Reply by davidkopp** [reply]
Ok, finally, my third version and probably also the last one: Dockerfile for only building and exporting.
I am now using this approach because I already have Nginx in use as a reverse proxy for other services on the same server, which now simply delivers the files as well. The purpose of the Dockerfile is that the complete build process takes place in a container and no dependencies need to be installed on the host system.

Dockerfile (`Dockerfile.export`):

```Dockerfile
FROM node:18 as base
WORKDIR /usr/src/app
COPY package*.json ./
RUN npm install
COPY . .

FROM base as builder
WORKDIR /usr/src/app
RUN npm run build

FROM scratch AS export
COPY --from=builder /usr/src/app/dist /
```

Build command that places the resulting `dist` directory in a `dist` directory on the host system:

```sh
podman build -f Dockerfile.export --output dist .
```

**Reply by wchorski** [reply]
Any advice on how to process markdown files in the vault before it hits the 11ty app repo? 
- Can we use the Obsidian Plugin outside of Obsidian?
- Do we have to use GitHub for this process?

**Reply by davidkopp** [reply]
@wchorski The Obsidian Digital Garden Plugin is only available for Obsidian and currently only supports GitHub.

To make some changes to your markdown files before they gets processed by 11ty you can use the feature "Custom Filters": https://dg-docs.ole.dev/getting-started/05-other-settings/#custom-filters
For example, I'm using custom filters to remove links to local files.

**Reply by janfromm** [reply]
Hi!
In the root folder of my digital garden repo there is no subfolder named netlify.
When trying to run app.js, I'm receiving an error:

```
node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module './netlify/functions/search/search.js'
Require stack:
- /[...]/self-host/app.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/[...]/self-host/app.js:5:23)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '[...]/self-host/app.js' ]
}

Node.js v20.11.0
```

What's my mistake?
Thanks for being pacient, I'm just getting started with node.js

Jan

**Reply by syx828** [reply]
@janfromm  This can work after version1.53.0

```
const express = require('express')
const app = express()
const port = 8080; 

app.get('/api/search', (req, res) => {
  res.status(200).send('Search is now handled on the client side with FlexSearch.');
})

app.use(express.static('dist'))

app.get('*', (req, res) => {
  res.redirect("/404")
})

app.listen(port, () => {
  console.log(`Digital garden running on port ${port}`)
})
```

**Reply by evlibra** [reply]
This app.js code worked for me:

const express = require('express')
const app = express()
const port = 8080;

app.get('/api/search', (req, res) => {
  res.status(200).send('Search is now handled on the client side with FlexSearch.');
})

app.use(express.static('dist'))

app.get('/{*any}', (req, res) => {
  res.redirect("/404")
})

app.listen(port, () => {
  console.log(`Digital garden running on port ${port}`)
})

For corrections of line 11, 12 thanks to https://stackoverflow.com/a/79680674/4218930

Also, had to perform:
nvm install 21
nvm use 21
npm install --include=optional sharp

**Comment by naquad** [comment]
Could you please clarify, who performs the actual Markdown -> HTML rendering?
The server script does not handle the notes files in any way. Is it some Vercel/Netlify magic?

Thank you.

**Comment by dayne** [comment]
@naquad the answer is [Eleventy](https://www.11ty.dev/).  Under the hood this plugin is taking the markdown files selected for publishing, bundling them up into a github repository, that is then rendered into a static website using the 11ty static site generator.  The rendering of the github repo has to be "run" somewhere, and the default pathway documented is to hook up the github repo to vercel for the rendering+hosting dance.   It could be done 100% on GitHub + GitHub Actions if you want to skip that part.

**Reply by naquad** [reply]
Thank you. I somehow thought it was UI-related rather than the whole rendering.

**Reply by ninijay** [reply]
can you elaborate how that could work with GitHub + GitHub Actions? I'm unable to find a guide

**Comment by rubenrivera** [comment]
@janfromm 

The template repository has changed since @oleeskild posted his answer. One significant change is that the default hosting platform was Netlify, but it was swapped to Vercel. I'm not sure if this swapping caused the removal of the Netlify directory, but new repositories don't include it.

Try commenting out or removing the statements that use `searchHandler`
```
const express = require('express')
const app = express()
const port = 8080; 
/**
const searchHandler = require('./netlify/functions/search/search.js').handler;

app.get('/api/search', async (req, res) => {
  //Mock the netlify event
  let event = {queryStringParameters: req.query};

  let response = await searchHandler(event);
  res.status(response.statusCode).send(response.body);
})
*/

app.use(express.static('dist'))

app.get('*', (req, res)=>{
  res.redirect("/404")
})

app.listen(port, () => {
  console.log(`Digital garden running on port ${port}`)
})
```

> Hi! In the root folder of my digital garden repo there is no subfolder named netlify. When trying to run app.js, I'm receiving an error:
> 
> ```
> node:internal/modules/cjs/loader:1147
>   throw err;
>   ^
> 
> Error: Cannot find module './netlify/functions/search/search.js'
> Require stack:
> - /[...]/self-host/app.js
>     at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
>     at Module._load (node:internal/modules/cjs/loader:985:27)
>     at Module.require (node:internal/modules/cjs/loader:1235:19)
>     at require (node:internal/modules/helpers:176:18)
>     at Object.<anonymous> (/[...]/self-host/app.js:5:23)
>     at Module._compile (node:internal/modules/cjs/loader:1376:14)
>     at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
>     at Module.load (node:internal/modules/cjs/loader:1207:32)
>     at Module._load (node:internal/modules/cjs/loader:1023:12)
>     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12) {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: [ '[...]/self-host/app.js' ]
> }
> 
> Node.js v20.11.0
> ```
> 
> What's my mistake? Thanks for being pacient, I'm just getting started with node.js
> 
> Jan

**Reply by janfromm** [reply]
@rubenrivera 

Hello,
thank you for your explanation, which I unfortunately only noticed recently.

I ended up doing the following:

1.) I created a Netlify project as described in the Digital-Garden-Docs (https://dg-docs.ole.dev/advanced/hosting-alternatives/).

2.) I asked Chat-GPT to create a shell script that automatically pulls the changes from my Digital Garden Github repo to my own server and then creates the "dist" directory with static html files via the Netlify CLI (https://docs.netlify.com/cli/get-started/).

3.) I use the following Express server to provide the Dist directory:

`const express = require('express');
const path = require('path');

const app = express();
const PORT = process.env.PORT || 3000;

app.use(express.static(path.join(__dirname, 'dist')));

app.listen(PORT, () => {
  console.log(`Server läuft auf http://localhost:${PORT}`);
});`


I'm relatively happy with this system because it's been fast and reliable so far. The **Express server also gives me great opportunities** to combine the static blog with a web app and implement additional functions.

However, it really bothers me that I don't know **how Netlify converts the Digital Garden repo into static HTML**, because my entire blog depends on this (closed source, third-party) program. If it ever stops working like it does now, I wouldn't know what to do.

Does anyone here know a way to do the conversion locally without relying on a third-party service?

**Comment by mounta11n** [comment]
I think the most convenient way is to deploy your digital garden in the very same git repo, but as a new branch and as a github page. this way you dont need netlify, vercel, a proxy server, docker or anything else. just one git repo and nothing more.
you have to name the forked repo `your_user_name.github.io`. here is an example where i did exactly this (with my other github account):

https://github.com/geschenkwald/geschenkwald.github.io

so the belongig digital-garden is (by github design, as mentioned, username.github.io): [geschenkwald.github.io/](https://geschenkwald.github.io/)

or, as you can see in my modified build.yaml file, you can set a custom CNAME, which I set to: [cognithink.de](https://cognithink.de)

Additionally what is really really cool, you can integrate [giscus](https://giscus.app/) in your blog/digital garden by utilizing the discussion function of the same repo as well. to get this working you have to edit the template files slightly (which are `src/site/_includes/layouts/index.njk` and `note.njk`). so in my case for example i basically added the following:

```
  <!-- ... -->
      {% endfor %}
  {{ content | hideDataview | taggify | link | safe}}
  
  <!-- comments section placeholder -->
  <div class="giscus"></div>
  
  <!-- giscus-script -->
  <script src="https://giscus.app/client.js"
          data-repo="geschenkwald/geschenkwald.github.io"
          data-repo-id="R_kgDOLBDuTA"
          data-category="Announcements"
          data-category-id="DIC_kwDOLBDuTM4Cf6VU"
          data-mapping="title"
          data-strict="0"
          data-reactions-enabled="0"
          data-emit-metadata="0"
          data-input-position="top"
          data-theme="light"
          data-lang="de"
          data-loading="lazy"
          crossorigin="anonymous"
          async>
  </script>
  
  {% for imp in dynamics.common.afterContent %}
    <!-- ... -->
```

And as you can see there is now a perfectly working comments sections below each of my blog articles and since i added myself (this account) as a collaborator in the settings of the geschenkwald.github.io repo, i can directly moderate, edit, delete comments in the discussion section without even having to logout and login with the other account.

@oleeskild if you are interested in this approach, i could make a new branch with a clean code and a pull request (or two, one for gh-page deployment and one for the giscus feature). additionally for giscus we could add an integration in the obsidian plugin, for example an option which the user can opt-in (like the file-tree-sidebar enable/disable etc).

**Reply by juanjesusov** [reply]
Hi! @mounta11n I consider that your proposal is just what I was looking for, I have done the steps you mention in [this repo](https://github.com/juanjesusov/juanjesusov.github.io). Then from obsidian I have sent the commits to generate the web, however the link associated to github pages displays the following screen which is basically the README file. 

I would like to ask if you know what step I am missing, I assume it is some configuration to define the path to the files I have generated with obsidian, however I have not been able to find the solution.

![image](https://github.com/oleeskild/obsidian-digital-garden/assets/78814508/1ae919f7-d3f6-4be1-adbe-cb35cfdb1355)

**Reply by pedroporo** [reply]
the github pages sould look like this 
![imagen](https://github.com/user-attachments/assets/ca7d3794-d9f5-49ac-908a-16b55751ebaa)
If you want you can look at my [repo](https://github.com/pedroporo/pedroporo.github.com) in case it helps you

**Reply by foxblock** [reply]
Many thanks for this @mounta11n 

I went ahead and created a fork of the template with the necessary changes, so anyone can use it as a starting point. This also includes a README with the setup steps:
https://github.com/foxblock/digitalgarden_gh-pages

**Comment by carakesa** [comment]
This is brilliant and basically what I was hoping to do when I first
started wanting to do a digital garden and the whole Vercel setup (and it's
associated bits and services) is really what turned me away.

I'll have to check this out and see if I can put it in practice on my end

**Reply by mounta11n** [reply]
le

**Comment by Ryj3l** [comment]
<img width="1285" alt="Screenshot 2024-09-08 at 11 08 05 AM" src="https://github.com/user-attachments/assets/ebe4744f-e723-4960-9072-fcf78611740b">
### Then write this Dockerfile

```# Stage 1: Build the application
FROM node:18 AS build
WORKDIR /usr/src/app

# Copy package files to install dependencies
COPY package*.json ./

# Install all dependencies (including devDependencies)
RUN npm install

# Copy the rest of the application code
COPY . .

# (Optional) Build the application if there's a build step, but make sure to skip it gracefully if it's missing
RUN if [ -f package.json ] && grep -q "build" package.json; then npm run build; else echo "No build step"; fi

# Stage 2: Production environment
FROM node:18-slim

# Set environment variables for production
ENV NODE_ENV=production

# Set the working directory
WORKDIR /usr/src/app

# Copy the package files (for production dependencies)
COPY package*.json ./

# Install only production dependencies
RUN npm ci --only=production

# Copy the application code and built files from the build stage
COPY --from=build /usr/src/app .

# Expose the port that the app will run on
EXPOSE 8080

# Start the application
CMD ["node", "app.js"]
```

### Nginx.conf

```server {
    listen 80;
    server_name yourdomain.com www.yourdomain.com;

    location /blog/ {
        proxy_pass http://localhost:8080/;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }

    return 301 https://$host$request_uri;
}

server {
    listen 443 ssl;
    server_name yourdomain.com www.yourdomain.com;

    ssl_certificate /etc/letsencrypt/live/yourdomain.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/yourdomain.com/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

    location /blog/ {
        proxy_pass http://localhost:8080/;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}
```

At this point as you know all you need is a docker-compose file. Build your stack with nginx.

**Comment by scrapix** [comment]
Hello, can someone please share this project as consumable docker image?
I'm not able to build on osx :?

**Reply by dangehub** [reply]
> Hello, can someone please share this project as consumable docker image? I'm not able to build on osx :?

I tried building one, but I only tested it in wsl.

https://hub.docker.com/r/qudange/obsidian-digital-garden

**Comment by foxblock** [comment]
Based on the work of @mounta11n (https://github.com/oleeskild/obsidian-digital-garden/discussions/160#discussioncomment-9694556), I created a fork of the template which uses GitHub pages as a method of deployment.
The setup is fairly easy and only the GitHub account and repository you need anyway are required (no Vercel, Docker, etc.).
As an additional bonus including comments using GitHub Discussions is fairly easy as well.

Use this template and follow the steps in the included README: https://github.com/foxblock/digitalgarden_gh-pages

**Comment by noahc3** [comment]
Since this thread is linked in the Digital Garden plugin docs, just want to drop here that I've created a Docker container and a helper script that lets you host your Digital Garden on any server/VPS/container/etc. you'd like: https://github.com/noahc3/dg-docker

It uses nginx for hosting the static content for performance with a small Express.js script to handle automatically pulling and rendering changes on commit to your Digital Garden content repo, triggered via GitHub webhook. I'm using it to host my content and it works great! It also does not require any changes to the actual contents of your Digital Garden repo so merging in updates to the template is no problem. 

If you have any suggestions feel free to open an issue or pull request on the repo.
